Can't mock dynamically imported packages

### Describe the bug

Scoped packages, e.g. `@wdio/logger` can't be mocked at all. In Jest it was possible to have the following structure:

```ts
__mocks__/
  - @wdio/
    - logger.ts
```

Jest would automatically pick this up.

### Reproduction

See https://stackblitz.com/edit/vitest-dev-vitest-tx14cb?file=package.json,test%2Fautomocking.spec.ts,__mocks__%2F%40wdio%2Flogger.ts&initialPath=__vitest__

I mocked out `@wdio/logger` in a way I expect Jest to work. The mocked file looks like this:
```ts
export default 'foobar';
```
In my test file (automocking.spec.ts) I do:
```ts
vi.mock('@wdio/test');
test('can automock scoped NPM package', () => {
  expect(foo).toBe('foobar');
});
```

The test failed due to:
```
can automock scoped NPM package
AssertionError: expected [Function getLogger] to be 'foobar' // Object.is equality
 - /test/automocking.spec.ts:38:41
``` 

### System Info

```shell
n/a
```


### Used Package Manager

yarn

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vite-node/src/client.ts
+++ b/packages/vite-node/src/client.ts
@@ -4,7 +4,7 @@ import vm from 'vm'
 import { dirname, extname, isAbsolute, resolve } from 'pathe'
 import { isNodeBuiltin } from 'mlly'
 import createDebug from 'debug'
-import { isPrimitive, normalizeModuleId, normalizeRequestId, slash, toFilePath } from './utils'
+import { isPrimitive, mergeSlashes, normalizeModuleId, normalizeRequestId, slash, toFilePath } from './utils'
 import type { ModuleCache, ViteNodeRunnerOptions } from './types'
 
 const debugExecute = createDebug('vite-node:client:execute')
@@ -135,7 +135,7 @@ export class ViteNodeRunner {
         if (importer && importer.startsWith('mock:'))
           importer = importer.slice(5)
         const { id } = await this.options.resolveId(dep, importer) || {}
-        dep = id && isAbsolute(id) ? `/@fs/${id}` : id || dep
+        dep = id && isAbsolute(id) ? mergeSlashes(`/@fs/${id}`) : id || dep
       }
 
       return dep

--- a/packages/vite-node/src/utils.ts
+++ b/packages/vite-node/src/utils.ts
@@ -8,6 +8,10 @@ export function slash(str: string) {
   return str.replace(/\\/g, '/')
 }
 
+export function mergeSlashes(str: string) {
+  return str.replace(/\/\//g, '/')
+}
+
 export function normalizeRequestId(id: string, base?: string): string {
   if (base && id.startsWith(base))
     id = `/${id.slice(base.length)}`

--- a/packages/vitest/src/runtime/mocker.ts
+++ b/packages/vitest/src/runtime/mocker.ts
@@ -100,7 +100,7 @@ export class VitestMocker {
   }
 
   public resolveDependency(dep: string) {
-    return normalizeRequestId(dep).replace(/^\/@fs\//, isWindows ? '' : '/')
+    return normalizeRequestId(dep.replace(this.root, '')).replace(/^\/@fs\//, isWindows ? '' : '/')
   }
 
   public normalizePath(path: string) {

--- a/test/core/src/dynamic-import.ts
+++ b/test/core/src/dynamic-import.ts
@@ -1,0 +1,4 @@
+export async function dynamicImport(name: string) {
+  const pkg = await import(name)
+  return pkg
+}

--- a/test/core/test/mock-internals.test.ts
+++ b/test/core/test/mock-internals.test.ts
@@ -2,6 +2,7 @@ import childProcess, { exec } from 'child_process'
 import timers from 'timers'
 import { expect, test, vi } from 'vitest'
 import { execDefault, execHelloWorld, execImportAll } from '../src/exec'
+import { dynamicImport } from '../src/dynamic-import'
 
 vi.mock('child_process')
 vi.mock('timers') // node built in inside __mocks__
@@ -19,4 +20,11 @@ test('node internal is mocked', () => {
 
 test('builtin is mocked with __mocks__ folder', () => {
   expect(timers.clearInterval()).toBe('foo')
+})
+
+test('mocked dynamically imported packages', async () => {
+  const mod = await dynamicImport('timers')
+  expect(mod).toHaveProperty('default')
+  expect(mod.default).toHaveProperty('clearInterval')
+  expect(mod.default.clearInterval()).toBe('foo')
 })

